ip_country/src/ip_country.rs

```diff
@@ -0,0 +1,276 @@
+use crate::bit_queue::BitQueue;
+use crate::country_block_serde::CountryBlockSerializer;
+use crate::country_block_stream::CountryBlock;
+use std::io;
+
+#[allow(unused_must_use)]
+pub fn ip_country(
+    _args: Vec<String>,
+    stdin: &mut dyn io::Read,
+    stdout: &mut dyn io::Write,
+    stderr: &mut dyn io::Write,
+) -> i32 {
+    let mut serializer = CountryBlockSerializer::new();
+    let mut line_number = 0usize;
+    let mut csv_rdr = csv::Reader::from_reader(stdin);
+    let mut errors = csv_rdr
+        .records()
+        .map(|string_record_result| match string_record_result {
+            Ok(string_record) => CountryBlock::try_from(string_record),
+            Err(e) => Err(format!("CSV format error: {:?}", e)),
+        })
+        .flat_map(|country_block_result| {
+            line_number += 1;
+            match country_block_result {
+                Ok(country_block) => {
+                    serializer.add(country_block);
+                    None
+                }
+                Err(e) => Some(format!("Line {}: {}", line_number, e)), // TODO no test for this line yet
+            }
+        })
+        .collect::<Vec<String>>();
+    let (ipv4_bit_queue, ipv6_bit_queue) = serializer.finish();
+    if let Err(error) = generate_rust_code(ipv4_bit_queue, ipv6_bit_queue, stdout) {
+        errors.push(format!("Error generating Rust code: {:?}", error)) // TODO no test for this line yet
+    }
+    if errors.is_empty() {
+        return 0;
+    } else {
+        let error_list = errors.join("\n");
+        write!(
+            stdout,
+            r#"
+            *** DO NOT USE THIS CODE ***
+            It will produce incorrect results.
+            The process that generated it found these errors:
+
+{}
+
+            Fix the errors and regenerate the code.
+            *** DO NOT USE THIS CODE ***
+"#,
+            error_list
+        );
+        write!(stderr, "{}", errors.join("\n"));
+        return 1;
+    }
+}
+
+fn generate_rust_code(
+    ipv4_bit_queue: BitQueue,
+    ipv6_bit_queue: BitQueue,
+    output: &mut dyn io::Write,
+) -> Result<(), io::Error> {
+    write!(output, "\n// GENERATED CODE: REGENERATE, DO NOT MODIFY!\n")?;
+    //TODO add number of country blocks to each run and create getters to retrieve number of blocks
+    generate_country_data("ipv4_country_data", ipv4_bit_queue, output)?;
+    generate_country_data("ipv6_country_data", ipv6_bit_queue, output)?;
+    Ok(())
+}
+
+fn generate_country_data(
+    name: &str,
+    mut bit_queue: BitQueue,
+    output: &mut dyn io::Write,
+) -> Result<(), io::Error> {
+    let bit_queue_len = bit_queue.len();
+    write!(output, "\n")?;
+    write!(output, "pub fn {}() -> (Vec<u64>, usize) {{\n", name)?;
+    write!(output, "    (\n")?;
+    write!(output, "        vec![")?;
+    let mut values_written = 0usize;
+    while bit_queue.len() >= 64 {
```

Comment:
You could make the 64 a manifest constant. Something like: COUNTRY_BLOCK_BIT_SIZE